### PR TITLE
feat(ngMessages): add support for default message

### LINKED ISSUE #12008: ngMessageDefault - Add support for showing values not mapped with ng-message
Many times it is necessary that we show a generic error message for unexpected values.
Would be nice to have something like:

``` HTML
<div ng-messages="controller.errors">
    <div ng-message="error1">Option 1</div>
    <div ng-message="error2">Option 2</div>
    <div ng-message-default>Option 3</div>
</div>
```

This way, whenever there is an "error3" in "controller.errors", as it is not mapped, the value under the ng-message-default would be shown.

### PR DESCRIPTION
**Pull request by Narretz**
Added support for showing default message when no values are mapped with ng-message.

Closes #12008

**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**

**What is the current behavior? (You can also link to an open issue here)**

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?**

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- [ ] Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated
- [ ] Fix/Feature: Tests have been added; existing tests pass

**Other information**:

### CODE REVIEW

File: src/ngMessages/messages.js [hunk 1]
```diff
@@ -428,23 +454,31 @@ angular.module('ngMessages', [], function initAngularHelpers() {
...
         };
 
-        this.deregister = function(comment) {
-          var key = comment.$$ngMessageNode;
-          delete comment.$$ngMessageNode;
-          removeMessageNode($element[0], comment, key);
-          delete messages[key];
+        this.deregister = function(comment, isDefault) {
+          if (isDefault) {
+            delete ctrl.default;
```
Comment (Narretz) [inline src/ngMessages/messages.js]: In the og PR you wondered if we need to detach the default controller first here @gkalpak - I've tested this and in defautl circumstances it definitely gets detached.

File: src/ngMessages/messages.js [hunk 2]
```diff
@@ -257,7 +257,21 @@ var jqLite;
  * .some-message.ng-leave.ng-leave-active {}
  * ```
  *
- * {@link ngAnimate Click here} to learn how to use JavaScript animations or to learn more about ngAnimate.
+ * {@link ngAnimate See the ngAnimate docs} to learn how to use JavaScript animations or to learn
+ * more about ngAnimate.
+ *
+ * ## Displaying a default message
+ * If the ngMessages renders no inner ngMessage directive (that is to say when the key values does not
```
Comment (petebacondarwin) [inline src/ngMessages/messages.js]: The wording here does not sound right. Should it be "(that is to say when no value matches any of the ngMessage directives)"?

**Review by petebacondarwin — APPROVED** [review]
LGTM with one question

File: src/ngMessages/messages.js (hunk 2 above)
Comment (Narretz) [reply src/ngMessages/messages.js]: Yeah, it's not correct. It should be something like "when a key in the ngMessages collection is truthy, but no matching message is defined"

**Comment by Narretz** [comment]
I just noticed that the behavior of the default message is not correct, at least not how it was requested in the original issue: #12008 
In this test, ngMessageDefault should not be shown because `col` doesn't have any keys:       https://github.com/angular/angular.js/pull/16587/files#diff-3e5c39c5b959703f8ff9efc61af2c103R665

However, that raises the question how ngMessageDefault should deal with multiple errors when one of them is defined and the other is not

File: src/ngMessages/messages.js [hunk 3]
```diff
@@ -409,8 +437,15 @@ angular.module('ngMessages', [], function initAngularHelpers() {
...
+            // Unset default message if set
+            if (ctrl.default) ctrl.default.detach();
+
             $animate.setClass($element, ACTIVE_CLASS, INACTIVE_CLASS);
           } else {
+
+            // Set default message if keys in collection do not match any message
+            if (ctrl.default && truthyKeys > 0) ctrl.default.attach();
+
             $animate.setClass($element, INACTIVE_CLASS, ACTIVE_CLASS);
```
Comment (petebacondarwin) [inline src/ngMessages/messages.js]: This line should only be called if there is no default...  Or perhaps this whole block should look like:

```js
var messageMatched = unmatchedMessages.length !== totalMessages;

if (ctrl.default) {
  if (messageMatched) {
    ctrl.default.detach();
  } else if (truthyKeys > 0) {
    ctrl.default.attach();
  }
}
if (messageMatched || ctrl.default) {
  $animate.setClass($element, ACTIVE_CLASS, INACTIVE_CLASS);
} else {
  $animate.setClass($element, INACTIVE_CLASS, ACTIVE_CLASS);
}
```

**Review by petebacondarwin — CHANGES_REQUESTED** [review]

File: test/ngMessages/messagesSpec.js [hunk 4]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-describe('ngMessages', function() {
+fdescribe('ngMessages', function() {
```
Comment (petebacondarwin) [inline test/ngMessages/messagesSpec.js]: Oops!

**Review by petebacondarwin — APPROVED** [review]
except for the `fdescribe`
